spec/app/ics-101-interchain-swap/README.md

```diff
@@ -0,0 +1,1635 @@
+  };
+
+
+  const sendPacketErr = await store.sendIBCSwapPacket(msg.port, msg.channel, timeoutHeight, timeoutStamp, packet);
+  abortTransactionUnless(sendPacketErr === undefined)
+
+  return {};
+}
+
+function singleAssetDeposit(msg: MsgSingleAssetDepositRequest): MsgSingleAssetDepositResponse {
+  // Validate message
+  const validationErr = msg.validateBasic();
+  abortTransactionUnless(validationErr === undefined);
+
+  // Check if pool exists
+  const { pool, found } = store.getInterchainLiquidityPool(msg.poolId);
+  abortTransactionUnless(found);
+
+  // Deposit token to escrow account
+  const balance = store.getBalance(msg.sender, msg.token.denom);
+  abortTransactionUnless(balance.amount.gt(sdk.NewInt(0)));
+
+  // Check pool status
+  abortTransactionUnless(pool.status === "ACTIVE");
+
+  // Lock tokens in escrow account
+  const lockErr = store.lockTokens(pool.counterPartyPort, pool.counterPartyChannel, msg.sender, sdk.NewCoins(msg.token));
+  abortTransactionUnless(lockErr === undefined);
+
+  const amm = new InterchainMarketMaker(pool);
+
+  const poolToken = await amm.depositSingleAsset(msg.token);
+  abortTransactionUnless(poolToken === undefined);
+
+  const packet: IBCSwapPacketData = {
+    type: "SINGLE_DEPOSIT",
+    data: protobuf.encode(msg);,
+    stateChange: { poolTokens: [poolToken] },
+  };
+
+  const sendPacketErr = await store.sendIBCSwapPacket(msg.port, msg.channel, timeoutHeight, timeoutStamp, packet);
+  abortTransactionUnless(sendPacketErr === undefined);
+
+  return { poolToken: pool.supply };
+}
+
+
+function multiAssetWithdraw(msg: MsgMultiAssetWithdrawRequest): MsgMultiAssetWithdrawResponse {
+  // Validate message
+  const validationErr = msg.validateBasic();
+  abortTransactionUnless(validationErr === undefined);
+
+  // Check if pool token denom exists
+  const poolTokenDenom = msg.poolToken.denom;
+  abortTransactionUnless(store.bankKeeper.hasSupply(poolTokenDenom));
+
+  // Get the liquidity pool
+  const { pool, found } = k.getInterchainLiquidityPool(ctx, poolTokenDenom);
+  abortTransactionUnless(found);
+
+  const amm = new InterchainMarketMaker(pool);
+
+  const outs = await amm.multiAssetWithdraw(msg.poolToken);
+  abortTransactionUnless(outs === undefined);
+
+  const packet: IBCSwapPacketData = {
+    type: "MULTI_WITHDRAW",
+    data: protobuf.encode(msg),
+    stateChange: {
+      out: outs,
+      poolTokens: [msg.poolToken],
+    },
+  };
+
+  const sendPacketErr = await store.sendIBCSwapPacket(ctx, msg.port, msg.channel, timeoutHeight, timeoutStamp, packet);
+  abortTransactionUnless(sendPacketErr === undefined);
+
+  return {};
+}
+
+
+function swap(msg: MsgSwapRequest): MsgSwapResponse {
+  const { pool, found } = store.getInterchainLiquidityPool(msg.poolId);
+  abortTransactionUnless(found);
+
+  abortTransactionUnless(pool.status === "ACTIVE");
+
+  const lockErr = store.lockTokens(pool.counterPartyPort, pool.counterPartyChannel, msg.sender, msg.tokenIn);
```

Comment:
Is this correct that tokens are locked when the user does both a left or a right swap? 

If I understand it correctly, in the case of a right swap, the user should specify how many tokens they want to buy, and the user doesn't know yet how much it needs to pay for them, right?

Reply:
They can know this. Market function include swap fee and then they can estimate output amount. 
In go implementation, Market functions is just stateless(don't update state variable directly and provide only estimation value). so before sending swap request, they can estimate how much they can get. 
Based on that, they can make swap request. 

Also based on slippage, they can avoid to fail in swap.